backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchKnnPredicate.java

```diff
@@ -0,0 +1,261 @@
+			if ( !vector.getClass().isArray() ) {
+				throw new IllegalArgumentException( "Vector can only be either a float or a byte array (float[], byte[])." );
+			}
+			if ( !vectorElementsType.equals( vector.getClass().getComponentType() ) ) {
+				throw log.vectorKnnMatchVectorTypeDiffersFromField( absoluteFieldPath, vectorElementsType,
+						vector.getClass().getComponentType() );
+			}
+			if ( Array.getLength( vector ) != indexedVectorsDimension ) {
+				throw log.vectorKnnMatchVectorDimensionDiffersFromField( absoluteFieldPath, indexedVectorsDimension,
+						Array.getLength( vector )
+				);
+			}
+			this.vector = vectorToJsonArray( vector, vectorElementsType );
+		}
+
+		@Override
+		public void filter(SearchPredicate filter) {
+			this.filter = ElasticsearchSearchPredicate.from( scope, filter );
+		}
+
+	}
+
+	private static JsonArray vectorToJsonArray(Object vector, Class<?> vectorElementsType) {
+		// we know it is an array since we've checked it when we got the vector
+		int length = Array.getLength( vector );
+		JsonArray array = new JsonArray( length );
+		for ( int i = 0; i < length; i++ ) {
+			if ( byte.class.equals( vectorElementsType ) ) {
+				array.add( Array.getByte( vector, i ) );
+			}
+			else {
+				array.add( Array.getFloat( vector, i ) );
+			}
+		}
+		return array;
+	}
+
+	private static class ElasticsearchImpl extends ElasticsearchKnnPredicate {
+
+		private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+		private static final JsonAccessor<String> FIELD_ACCESSOR = JsonAccessor.root().property( "field" ).asString();
+		private static final JsonArrayAccessor QUERY_VECTOR_ACCESSOR = JsonAccessor.root().property( "query_vector" ).asArray();
+		private static final JsonAccessor<Integer> K_ACCESSOR = JsonAccessor.root().property( "k" ).asInteger();
+
+		private static final JsonObjectAccessor FILTER_ACCESSOR = JsonAccessor.root().property( "filter" ).asObject();
+		private static final JsonAccessor<Integer> NUM_CANDIDATES_ACCESSOR =
+				JsonAccessor.root().property( "num_candidates" ).asInteger();
+
+
+		private ElasticsearchImpl(Builder<?> builder) {
+			super( builder );
+		}
+
+		@Override
+		protected JsonObject doToJsonKnn(PredicateRequestContext context) {
+			JsonObject object = new JsonObject();
+			FIELD_ACCESSOR.set( object, absoluteFieldPath );
+			K_ACCESSOR.set( object, k );
+			if ( filter != null ) {
+				FILTER_ACCESSOR.set( object, filter.toJsonQuery( context ) );
+			}
+			// TODO: this is a required parameter in Elasticsearch:
+			//    {
+			//      "error": {
+			//        "root_cause": [
+			//          {
+			//            "type": "illegal_argument_exception",
+			//            "reason": "Required [num_candidates]"
+			//          }
+			//        ],
+			//        "type": "illegal_argument_exception",
+			//        "reason": "Required [num_candidates]"
+			//      },
+			//      "status": 400
+			//    }
+			NUM_CANDIDATES_ACCESSOR.set( object, 100 );
```

Comment:
this is a required filed  for a knn search in Elasticsearch and currently we do not read it. so I was wondering how should we do it... just add it as an optional property to the predicate or create some predicate that is only available through an extension ... (NOTE that OpenSearch does not need this attribute)

Reply:
> and currently we do not read it

What do you mean? Read from where?

Reply:
I meant we do not expose a way to pass it from the search predicate (or through any other way) 😃. So should I just add it to `KnnPredicateOptionsStep` like this:
```java
public interface KnnPredicateOptionsStep extends PredicateScoreStep<KnnPredicateOptionsStep>, PredicateFinalStep {

	KnnPredicateOptionsStep filter(SearchPredicate searchPredicate);

	default KnnPredicateOptionsStep filter(PredicateFinalStep searchPredicate) {
		return filter( searchPredicate.toPredicate() );
	}

	KnnPredicateOptionsStep filter(Function<? super SearchPredicateFactory, ? extends PredicateFinalStep> clauseContributor);

	/**
	 * throws exceptions on OpenSearch and Lucene
	 */
	KnnPredicateOptionsStep numberOfCandidates(int numberOfCandidates);

}
```

Reply:
Doesn't look like a great solution if it doesn't make any sense on OpenSearch/Lucene... I need to understand what the point of this parameter is. I'll have a look.

Reply:
Here's a link if you haven't looked at it already:
https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#tune-approximate-knn-for-speed-accuracy

I was thinking about setting it to something like `k/num-of-shards+1` 😃 in the background ..